Sephora-M/assignment3/LetterGrid.java

```diff
@@ -0,0 +1,117 @@
+ * all the words that can be formed in the grid. 
+ * 
+ * @author Sephora-M
+ */
+public class LetterGrid {
+	char[][] grid;
+	int X;
+	int Y;
+	
+	public LetterGrid(char[][] grid) {
+		if (grid == null || grid[0].length == 0)
+			throw new IllegalArgumentException("Cannot instatiate an empty or a null grid.");
+		this.grid = grid;
+		Y = grid.length;
+		X = grid[0].length;
+	}
+	
+	/**
+	 * Given a Trie object, returns a set of all the words in the Trie that can be formed in the grid 
+	 * @param dict a Trie, the dictionary
+	 * @return set of all the words from the dictionary that can be from in the grid
+	 * 
+	 * Note: the idea is to start from each letter in the grid and perform a DFS-like search. To optimize
+	 * the search, check at each time if the current word that we are constructing is a prefix of a word in the 
+	 * dictionary. It is not exactly a DFS, the subtility here is that we don't just want to visit all the 
+	 * nodes, but we want to visit them from different path. This is taken care here by keeping track of a
+	 * depthMap, which tells us when we visited each node. 
+	 */
+	public Set<String> dictionnaryWords(Trie dict){
+		HashSet<String> foundWords = new HashSet<String>();
+
+		for(int x = 0; x < X; x++) {
+			for(int y = 0; y < Y; y++) {
+				int[][] depthMap = new int[Y][X];
+				int currentDepth = 1;
+				depthMap[y][x] = currentDepth;
+				Point firstLetterCoord = new Point(x,y);
+				String currentPrefix = getLetter(firstLetterCoord)+"";
+				if (dict.isWord(currentPrefix))
+					foundWords.add(currentPrefix);
+				if (dict.isPrefix(currentPrefix))
+					searchWords(firstLetterCoord, currentDepth, depthMap, dict, foundWords, currentPrefix);
+			}
+		}
+		
+		return foundWords;
+	}
+	
+    private void searchWords(Point letter, int currentDepth, int[][] depthMap, Trie dict,
+			HashSet<String> foundWords, String currentPrefix) {
+    	
+    	ArrayList<Point> neighbors = findNeighbors(letter, currentDepth, depthMap);
+    	
+    	if (neighbors.size() == 0)
+    		return;
+    	for (Point n : neighbors) {
+    		String newPrefix = currentPrefix + getLetter(n);
+    		if (dict.isPrefix(newPrefix)){
+    			int[][] newDepthMap = deepCopy(depthMap);
+    			newDepthMap[n.y][n.x] = currentDepth + 1;
+    			if (dict.isWord(newPrefix))
+    				foundWords.add(newPrefix);
+    			searchWords(n, currentDepth +1, newDepthMap, dict, foundWords, newPrefix);
+    		}
+    	}
+		
+	}
+    
+    private char getLetter(Point p) {
+    	return grid[p.y][p.x];
+    }
+    
+    private int[][] deepCopy(int[][] depthMap) {
+        int[][] copy = new int[Y][X];
+        
+        for (int x = 0; x < X; x++) {
+          for (int y = 0; y < Y; y++) {
+        		  copy[y][x] = depthMap[y][x];
+          }
+        }
+        
+        return copy;
+      }
+    
+	private ArrayList<Point> findNeighbors(Point p, int currentDepth, int[][] depthMap){
```

Comment:
Very nice! How about writing a few unit tests for this method? When I was writing my solution to this problem I actually had a bug in this method and couldn't find it for good 10 minutes. :) It's important to unit test methods that are either algorithmically complex or easy to do a typo in (accidentally writing a '+' instead of a '-' for example).